### PR TITLE
🐛 Ban use of the implicit global this in functions

### PR DESCRIPTION
**Pull request by jridgewell**
The underlying bug is that in strict mode, the implicit `this` of a
function is `undefined`. In sloppy, it was the global `window`.

This extends the work done in #30174 and #30235.

### CODE REVIEW

File: build-system/eslint-rules/no-invalid-this.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+ * limitations under the License.
+ */
+'use strict';
+
+// Disables use of the `this` value when we suspect that it is using the
+// implict global `this`. `this` is only valid inside class
+// methods/constructors, and nested arrow functions.
+//
+// We special case calls to `fn.bind`, `fn.call`, and `fn.apply`, since they
+// are usually forwarding the `this` context and not direclty using it. The
```
Comment (samouri) [inline build-system/eslint-rules/no-invalid-this.js:23]: Nit: typo with directly

File: build-system/eslint-rules/no-invalid-this.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+              if (id && /^[A-Z]/.test(id.name)) {
+                return;
+              }
+
+            case 'Program':
+              return context.report({
+                node,
+                message: 'invalid this',
+                fix(fixer) {
+                  return fixer.replaceText(node, 'undefined');
```
Comment (samouri) [inline build-system/eslint-rules/no-invalid-this.js:103]: Isn't the correct fix `globalThis`?

Comment (jridgewell) [reply build-system/eslint-rules/no-invalid-this.js:103]: Only if they intended to use the global this implicitly. I think most cases are errors, and this is what the normal strict mode would do.

File: extensions/amp-playbuzz/0.1/test/test-amp-playbuzz.js [hunk 3]
```diff
@@ -75,6 +75,8 @@ describes.realWin(
       ins.setAttribute('alt', 'Testing');
       if (opt_responsive) {
         ins.setAttribute('layout', 'responsive');
+      } else {
```
Comment (kristoferbaxter) [inline extensions/amp-playbuzz/0.1/test/test-amp-playbuzz.js:78]: Is this change related? I'm not following it regarding strict mode and missing `this`?

File: src/polyfills/array-includes.js [hunk 4]
```diff
@@ -24,9 +24,11 @@
  */
 function includes(value, opt_fromIndex) {
   const fromIndex = opt_fromIndex || 0;
+  // eslint-disable-next-line local/no-invalid-this
   const len = this.length;
   let i = fromIndex >= 0 ? fromIndex : Math.max(len + fromIndex, 0);
   for (; i < len; i++) {
+    // eslint-disable-next-line local/no-invalid-this
```
Comment (kristoferbaxter) [inline src/polyfills/array-includes.js:31]: Question: Given the failure case is broken binary, should this exclusion require a REVIEW or OK flag too?

Might not be necessary, but just wanted to ask for everyone's clarification.

**Review by kristoferbaxter — APPROVED** [review]
Not sure about the fixer, but other parts appear good with a few outstanding questions.

File: build-system/eslint-rules/no-invalid-this.js (hunk 2 above)
Comment (samouri) [reply build-system/eslint-rules/no-invalid-this.js:103]: In that case it probably makes more sense to have a human resolve

**Review by samouri — APPROVED** [review]
Lgtm. I'll let you decide what to do about the fixer
